[WIP] Add blog article about minReadySeconds for StatefulSets and maxSurge for DaemonSets (#35440)
kubernetes/website: content/en/blog/_posts/2022-07-27-sig-apps-GA-1.25/index.md

---
layout: blog
title: "sig-apps features graduating to stable in 1.25"
date: 2022-07-27
slug: "sig-apps features graduating to stable in 1.25"
---

> **Authors:** Ravi Gudimetla (Apple), Filip Krepensky (Red Hat), Maciej Szulik (Red Hat)

Reviewer: ```suggestion
**Authors:** Ravi Gudimetla (Apple), Filip Krepinsky (Red Hat), Maciej Szulik (Red Hat)
```

> This blog describes the two features namely `minReadySeconds for StatefulSets` and `maxSurge for Daemonsets` that sig-apps is happy to graduate to stable in 1.25

Reviewer: ```suggestion
This blog describes the two features namely `minReadySeconds for StatefulSets` and `maxSurge for DaemonSets` that sig-apps is happy to graduate to stable in 1.25
```

## What problems does these features solve?

### MinReadySeconds for StatefulSets

> `minReadySeconds` ensures that the statefulset workload is `Ready` for the given number of seconds before calling the

Reviewer 2: Keep the resource names always upper case:
```suggestion
`minReadySeconds` ensures that the StatefulSet workload is `Ready` for the given number of seconds before calling the
```

> pod `Available`. The notion of being `Ready` and `Available` is quiet important for workloads. For example, some workloads like Prometheus with multiple instances of Alertmanager should be considered `Available` only when the Alertmanager's state transfer is complete. `minReadySeconds` also helps when using loadbalancers with cloud providers. Since the pod should be `Ready` for the given number of seconds, it provides buffer time to prevent killing pods in rotation before new pods show up.

Reviewer: ```suggestion
pod `Available`. The notion of being `Ready` and `Available` is quite important for workloads. For example, some workloads like Prometheus with multiple instances of Alertmanager should be considered `Available` only when the Alertmanager's state transfer is complete. `minReadySeconds` also helps when using loadbalancers with cloud providers. Since the pod should be `Ready` for the given number of seconds, it provides buffer time to prevent killing pods in rotation before new pods show up.
```

Reviewer: I know the last sentence says it, but maybe we should mention specifically rollouts? We discussed something similar in the docs PR: https://github.com/kubernetes/website/pull/35539#discussion_r933175664

### MaxSurge for DaemonSets

> `MaxSurge` allows a daemonset workload to run multiple instances of the same pod on a node during rollout to minimize the downtime of the daemonset to other consumers. Kubernetes system-level components like CNI, CSI are typically run as daemonsets. These components can have impact on the availablity of the workloads if those daemonsets go down momentarily during the upgrades. The feature allows daemonset pods to surge, there by ensuring zero-downtime for the daemonsets.

Reviewer: ```suggestion
`MaxSurge` allows a daemonset workload to run multiple instances of the same pod on a node during rollout to minimize the downtime of the daemonset to other consumers. Kubernetes system-level components like CNI, CSI are typically run as daemonsets. These components can have impact on the availablity of the workloads if those daemonsets go down momentarily during the upgrades. The feature allows daemonset pods to surge, thereby ensuring zero-downtime for the daemonsets.
```

> `MaxSurge` allows a daemonset workload to run multiple instances of the same pod on a node during rollout to minimize the downtime of the daemonset to other consumers. Kubernetes system-level components like CNI, CSI are typically run as daemonsets. These components can have impact on the availablity of the workloads if those daemonsets go down momentarily during the upgrades. The feature allows daemonset pods to surge, there by ensuring zero-downtime for the daemonsets.

Reviewer 2: ```suggestion
`MaxSurge` allows a DaemonSet workload to run multiple instances of the same pod on a node during rollout to minimize the downtime of the DaemonSet to other consumers. Kubernetes system-level components like CNI, CSI are typically run as DaemonSets. These components can have impact on the availability of the workloads if those DaemonSets go down momentarily during the upgrades. The feature allows DaemonSet pods to surge, there by ensuring zero-downtime for the DaemonSets.
```

> Please note that the usage of `HostPort` in conjunction with `MaxSurge` in daemonsets is not allowed as daemonset pods are tied to a single node and two active pods cannot share the same port on the same node.

Reviewer 2: ```suggestion
Please note that the usage of `HostPort` in conjunction with `MaxSurge` in DaemonSets is not allowed as DaemonSet pods are tied to a single node and two active pods cannot share the same port on the same node.
```

Reviewer 2: And through the rest of the docs as well...

## How does it work?

### MinReadySeconds for StatefulSets

The statefulSet controller watches for the statefulset pods and counts how long a particular pod has been in the `Running` state, if this value is greater than or equal to the time specified in `.spec.minReadySeconds` field of the statefulset, the statefulset controller updates the `AvailableReplicas` field in the statefulset's status.

### MaxSurge for DaemonSets

> The `DaemonSet` controller creates the additional pods based on the value given in `.spec.strategy.rollingUpdate.maxSurge`. The additional pods would run on the same node where the old daemonset pod is running till the old pod gets killed. This value cannot be `0` when `MaxUnavailable` is 0. The default value is 0.

Reviewer 2: ```suggestion
The DaemonSet controller creates the additional pods (above the desired number resulting from DaemonSet spec) based on the value given in `.spec.strategy.rollingUpdate.maxSurge`. The additional pods would run on the same node where the old daemonset pod is running till the old pod gets killed. This value cannot be `0` when `MaxUnavailable` is 0. The default value is 0.
```
or something similar where we'd explicitly call out the fact that `.maxSurge` is above the usual replicas.

## How do I use it?

### MinReadySeconds for StatefulSets

> You are required to download and install a kubectl greater than v1.22.0 version

Reviewer 2: Irrelevant, I'd drop it.

> Specify a value for `minReadySeconds` for any StatefulSet and you check if pods are available or not by checking

Reviewer: ```suggestion
Specify a value for `minReadySeconds` for any StatefulSet and check if pods are available or not by inspecting
```

`AvailableReplicas` field using:

`kubectl get statefulset/<name_of_the_statefulset> -o yaml

Please note that the default value of `minReadySeconds` is 0

### MaxSurge for DaemonSets

Specify the update strategy to `RollingUpdate` and set `.spec.updateStrategy.rollingUpdate.maxSurge`

> 

Reviewer: `and observe a faster rollout and higher number of pods running at the same time in the next rollout`

what about adding the observe part as well in some form?
